### ISSUE TITLE
Performance of /oauth/token endpoint

### ISSUE THREAD

**Issue by KowalczykBartek** [cloudfoundry/uaa#936]
Hi UAA team :) 

### What version of UAA are you running?
version: 4.20.0

### How are you deploying the UAA?
locally only using gradlew  (osx, 2,6 GHz Intel Core i7)

###  Preface
Because I wanted to check what is UAA's limit of token generation, I ran small performance tests, simple wrk call on /oauth/token - nothing fancy. Result was not splendid so I checked for what UAA spend its CPU time, I attached async-profiler and this is the output 

<img width="1021" alt="screen shot 2019-01-02 at 20 35 24" src="https://user-images.githubusercontent.com/4283964/50608908-4f6b6180-0ece-11e9-8fbd-467b6df47a4b.png">

UAA, is spending more than 50% of time in **org.cloudfoundry.identity.uaa.zone.MultitenantJdbcClientDetailsService.loadClientByClientId**

I made small experiment and I added simple "cache" to that class (just ConcurrentHashMap) that holds result of 
```java
ClientDetails details;
try {
    details = jdbcTemplate.queryForObject(selectClientDetailsSql, new ClientDetailsRowMapper(), clientId, zoneId);
} catch (EmptyResultDataAccessException e) {
    throw new NoSuchClientException("No client with requested id: " + clientId);
}
```
and with that piece of code included, UAA behaves a bit better on that path
<img width="1114" alt="screen shot 2019-01-02 at 20 42 30" src="https://user-images.githubusercontent.com/4283964/50609132-1ed7f780-0ecf-11e9-9424-1161ad87ff22.png">
only key signing (here 2048bits) and some fasterxml stuff.

here, I uploaded original SVG files (github don't allow for uploading svg files in issue)
before: https://github.com/KowalczykBartek/tmp-svg/blob/master/slow.svg
after: https://github.com/KowalczykBartek/tmp-svg/blob/master/slow-but-better.svg

###  Question
Is it possible to add some caching on MultitenantJdbcClientDetailsService layer ? Does it makes sense from security point of view (or just any point of view :D ) to cache ClientDetails in memory (it can be cached on each instance without any sharing, like Redis or Memcached) ?

thanks !

**Comment by mwdb** [comment]
We have been looking into the topic of caching also.
Our observations are:

- Issuing a client credential token requires 14 calls to the data base, touching the identity_zone and oauth_client_details tables. Details are:
4 calls for identity_zone by id
2 calls for identity_zone by subdomain
8 calls to oauth_client_details by client_id and identity_zone_id

The simplest cache would be to have a request level cache, which avoids redundant calls for the same data. This would already take it from 14 calls  to the database to 3.

- For user based tokens like the password grant token numbers are worse: roughly 54 calls.
Details are:
20 calls to group_membership
10 calls to oauth_client_details
9 calls to users
15 calls to identity_zone

Similar, with a request level cache this could be reduced to ~ 6 calls.


Caching can be implemented using the Spring cache mechanisms. Concerning the cache implementations, I see the following requirements:
- The cache returns a copy of the cached object to avoid modifications of cached objects
- Within processing a single request, changes to persistent data can be ignored

**Comment by shamus** [comment]
Closing due to inactivity.

**Comment by mwdb** [comment]
@shamus 
This is an issue in UAA that deserves to be fixed.
Instead of two DB requests, UAA is at 14. This considerably costs performance and limits scaling.

Requesting to reopen it.

**Comment by strehle** [comment]
@mwdb correct, however I think someone should provide a PR so that we can discuss this

**Comment by KowalczykBartek** [comment]
I can try to do some PR, but first I wanted (at least I wanted in Jan 2019 xd) to get opinion about this ticket from core team :)

**Comment by shamus** [comment]
Oh hi @mwdb, we did a purge of all issues older than ~9 months that had no activity -- regardless of whether or not the issue was valid. We assumed we'd get feedback on important issues.

Thank you for your work (and the nice flame graphs) showing how the UAA bottlenecks on DB access for `client_credential` grants @KowalczykBartek. And thanks for the further breakdown @mwdb.

This does seem like a good topic to dig into.

What I'm not certain about is if using spring caching at the request level is the right solution to the problem. Were other approaches considered?

I have the impression that there are a set of beans scattered across the codebase that need access to data and have no other way to get it other than by querying. In that scenario, I understand why caching is attractive. But, I wonder if itis possible to reorganize these beans so that the can collaborate more effectively? Can the responsibility of retrieving data be consolidated into certain beans, while other beans simply accept the data for processing?

**Comment by mwdb** [comment]
@shamus For some beans this might be possible, e.g. for the identity zone information. This is anyways stored as thread local variable. For other beans like the OAuth client, it is more difficult as it is used at different places, e.g. 
- during authentication
- during token issuing 
Changing that would imply larger changes.

Personally I'd vote for changing persistence from jdbctemplate to Hibernate and open a Hibernate session per request. With this all requests first hit the Hibernate cache. Additionally Hibernates abstracts the different DB SQL flavours.
We'd volunteer for providing a PR. As persistence access is abstracted through interfaces, the change could be entity be entity to avoid a big-bang. So start with the clients, then the identity zone, etc.

To put some numbers around it: we spent as many CPU cores for the DB as we do for the UAA instances. For an authorization server this is to much. Many outages we have experienced  are caused by a DB overload.
